src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Shell/VSShellServices.cs

```diff
@@ -11,65 +12,54 @@ namespace Microsoft.VisualStudio.Shell;
 [AppliesTo(ProjectCapabilities.AlwaysApplicable)]
 internal class VSShellServices : IVsShellServices
 {
-
-    private readonly AsyncLazy<(bool IsCommandLineMode, bool IsPopulateSolutionCacheMode)> _initialization;
+    private static readonly object s_lockObject = new();
+    private static Task<bool>? s_isCommandLine;
+    private static Task<bool>? s_isPopulateCache;
 
     [ImportingConstructor]
     public VSShellServices(
-        IVsUIService<SVsShell, IVsShell> vsShellService,
-        IVsUIService<SVsAppCommandLine, IVsAppCommandLine> commandLineService,
-        JoinableTaskContext joinableTaskContext)
+        IVsService<IVsAppId> vsAppId,
+        IVsService<SVsAppCommandLine, IVsAppCommandLine> commandLine)
     {
-        _initialization = new(
-            async () =>
-            {
-                // Initialisation must occur on the main thread.
-                await joinableTaskContext.Factory.SwitchToMainThreadAsync();
-
-                IVsShell? vsShell = vsShellService.Value;
-                IVsAppCommandLine? commandLine = commandLineService.Value;
-
-                bool isCommandLineMode = IsCommandLineMode();
+        lock (s_lockObject)
+        {
+            s_isCommandLine ??= IsCommandLineModeAsync(vsAppId);
+            s_isPopulateCache ??= IsPopulateSolutionCacheModeAsync(commandLine);
 
-                if (isCommandLineMode)
-                {
-                    return (IsCommandLineMode: true, IsPopulateSolutionCacheMode());
-                }
+        }
 
-                return (false, false);
-
-                bool IsCommandLineMode()
-                {
-                    Assumes.Present(vsShell);
-
-                    int hr = vsShell.GetProperty((int)__VSSPROPID.VSSPROPID_IsInCommandLineMode, out object result);
+        static async Task<bool> IsCommandLineModeAsync(IVsService<IVsAppId> vsAppIdService)
+        {
+            IVsAppId? vsAppId = await vsAppIdService.GetValueOrNullAsync();
+            if (vsAppId is not null)
+            {
+                const int VSAPROPID_IsInCommandLineMode = (int)Microsoft.Internal.VisualStudio.AppId.Interop.__VSAPROPID10.VSAPROPID_IsInCommandLineMode;
+                return ErrorHandler.Succeeded(vsAppId.GetProperty(VSAPROPID_IsInCommandLineMode, out object o)) && o is bool isInCommandLineMode && isInCommandLineMode;
```

Comment:
nit: This is equivalent, but slightly more succinct:

```suggestion
                return ErrorHandler.Succeeded(vsAppId.GetProperty(VSAPROPID_IsInCommandLineMode, out object o)) && o is true;
```

Reply:
Changed. Definitely more succinct.